sqs/src/main/scala/akka/stream/alpakka/sqs/SqsBatchFlowStage.scala

```diff
@@ -0,0 +1,157 @@
+            sqsClient.sendMessageBatchAsync(
+              batch,
+              new AsyncHandler[SendMessageBatchRequest, SendMessageBatchResult] {
+                override def onError(exception: Exception): Unit = {
+                  val batchException = BatchException(messages.size, exception)
+                  responsePromise.failure(batchException)
+                  failureCallback.invoke(batchException)
+                }
+
+                override def onSuccess(request: SendMessageBatchRequest, result: SendMessageBatchResult): Unit =
+                  if (!result.getFailed.isEmpty) {
+                    val nrOfFailedMessages: Int = result.getFailed.size()
+                    val batchException: BatchException =
+                      BatchException(
+                        batchSize = messages.length,
+                        cause = new Exception(
+                          s"Some messages are failed to send. $nrOfFailedMessages of $nrOfMessages messages are failed"
+                        )
+                      )
+                    responsePromise.failure(batchException)
+                    failureCallback.invoke(batchException)
+                  } else {
+                    val results = ListBuffer.empty[Result]
+
+                    val successfulMessages = result.getSuccessful.iterator()
+                    while (successfulMessages.hasNext) {
+                      val successfulMessage: SendMessageBatchResultEntry = successfulMessages.next()
+                      val messageBody: String = messages(successfulMessage.getId.toInt)
+
+                      val sendMessageResult: SendMessageResult = new SendMessageResult()
+                        .withMD5OfMessageAttributes(successfulMessage.getMD5OfMessageAttributes)
+                        .withMD5OfMessageBody(successfulMessage.getMD5OfMessageBody)
+                        .withMessageId(successfulMessage.getMessageId)
+                        .withSequenceNumber(successfulMessage.getSequenceNumber)
+
+                      results += Result(sendMessageResult, messageBody)
+                    }
+
+                    responsePromise.success(results.toList)
+                    sendCallback.invoke(result)
+                  }
+              }
+            )
+            push(out, responsePromise.future)
+          }
+
+          private def createMessageBatch(messages: Array[String]): SendMessageBatchRequest = {
+            val messageRequestEntries: util.List[SendMessageBatchRequestEntry] =
+              new util.ArrayList[SendMessageBatchRequestEntry]()
+            var id = 0
+            messages.foreach { message =>
+              messageRequestEntries.add(new SendMessageBatchRequestEntry(id.toString, message))
+              id += 1
+            }
+
+            val batchRequest = new SendMessageBatchRequest(queueUrl, messageRequestEntries)
+            batchRequest
+          }
```

Comment:
@SergeiShabanau, the reason why I do not use `zipWithIndex` with `map` is to avoid unnecessary memory allocations and also please correct me if I am wrong but basically this code is iterating three times, but with one local variable I guess we can do that with one iteration. Even though according to Asymptotic Notation they are both O(n), I thought fewer iterations might be better. However, if it confuses the logic here, I can change that part.